File: apps/lantern-api/src/features/metric_showcase/application/emit_all_metric_types.py
import time
from src.features.telemetry.infrastructure.start_otel import Telemetry

PREFIX = "metric.showcase"


def emit_all_metric_types(telemetry: Telemetry) -> dict[str, str]:
    started = time.perf_counter()
    telemetry.up_down(f"{PREFIX}.inflight", 1)
    try:
        telemetry.increment(f"{PREFIX}.requests")
        latency_ms = (time.perf_counter() - started) * 1000.0
        telemetry.record(f"{PREFIX}.latency_ms", latency_ms)
        telemetry.set_gauge(f"{PREFIX}.queue_depth", float(int(latency_ms) % 10 + 1))
        return {
            "counter": f"{PREFIX}.requests",
            "histogram": f"{PREFIX}.latency_ms",
            "upDown": f"{PREFIX}.inflight",
            "gauge": f"{PREFIX}.queue_depth",
        }
    finally:
        telemetry.up_down(f"{PREFIX}.inflight", -1)

File: apps/atlas-api/src/features/metric-showcase/application/emit-all-metric-types.ts
import type { Telemetry } from "../../../telemetry/application/ports/telemetry";

const PREFIX = "metric.showcase";

/** Emit one sample of each OTLP instrument kind for guardrail/adoption demos. */
export function emitAllMetricTypes(telemetry: Telemetry): {
  counter: string;
  histogram: string;
  upDown: string;
  gauge: string;
} {
  const started = performance.now();
  telemetry.upDown(`${PREFIX}.inflight`, 1, { surface: "ts-server" });
  try {
    telemetry.increment(`${PREFIX}.requests`, { surface: "ts-server" });
    const latency = performance.now() - started;
    telemetry.record(`${PREFIX}.latency_ms`, latency, { surface: "ts-server" });
    telemetry.setGauge(`${PREFIX}.queue_depth`, Math.round(latency % 10) + 1, {
      surface: "ts-server",
    });
    return {
      counter: `${PREFIX}.requests`,
      histogram: `${PREFIX}.latency_ms`,
      upDown: `${PREFIX}.inflight`,
      gauge: `${PREFIX}.queue_depth`,
    };
  } finally {
    telemetry.upDown(`${PREFIX}.inflight`, -1, { surface: "ts-server" });
  }
}
